C error when smart casting interface to type

### Describe the bug

Trying to compile
```v
module main

interface Abc {
	a int
}

struct Xyz {
	a int
}

fn main() {
	d := Abc(Xyz{})
	if d is Xyz {
		println(d)
	}
}
```
gives
```
/tmp/v_1002/src.6834607243794453707.tmp.c:12663: error: cannot convert 'struct main__Xyz' to 'void *'
```

### Reproduction Steps

try to compile the above code

### Expected Behavior

compile successfully

### Current Behavior

```
/tmp/v_1002/src.6834607243794453707.tmp.c:12663: error: cannot convert 'struct main__Xyz' to 'void *'
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.2 bd52a94

### Environment details (OS name and version, etc.)

```yaml
V full version: V 0.4.2 4bc9a8f.bd52a94
OS: linux, Linux version 6.5.9-200.fc38.x86_64 (mockbuild@1cc7d6d790524511abd343182a21832a) (gcc (GCC) 13.2.1 20231011 (Red Hat 13.2.1-4), GNU ld version 2.39-15.fc38) #1 SMP PREEMPT_DYNAMIC Wed Oct 25 20:40:49 UTC 2023
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-4310U CPU @ 2.00GHz

getwd: /home/user/Projects/v-testing/webview
vexe: /home/user/.local/lib/v/v
vexe mtime: 2023-11-06 17:14:03

vroot: OK, value: /home/user/.local/lib/v
VMODULES: OK, value: /home/user/.vmodules
VTMP: OK, value: /tmp/v_1002

Git version: git version 2.41.0
Git vroot status: weekly.2023.44-67-gbd52a94e
.git/config present: true

CC version: cc (GCC) 13.2.1 20231011 (Red Hat 13.2.1-4)
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -113,12 +113,17 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 		if is_ptr && !is_var_mut {
 			ref_str := '&'.repeat(typ.nr_muls())
 			g.write('str_intp(1, _MOV((StrIntpData[]){{_SLIT("${ref_str}"), ${si_s_code} ,{.d_s = isnil(')
-			if is_ptr && typ.has_flag(.option) {
+			if typ.has_flag(.option) {
 				g.write('*(${g.base_type(exp_typ)}*)&')
 				g.expr(expr)
 				g.write('.data')
 				g.write(') ? _SLIT("Option(&nil)") : ')
 			} else {
+				inside_casting_to_str_old := g.inside_casting_to_str
+				g.inside_casting_to_str = false
+				defer {
+					g.inside_casting_to_str = inside_casting_to_str_old
+				}
 				g.expr(expr)
 				g.write(') ? _SLIT("nil") : ')
 			}

--- a/vlib/v/tests/interface_str_method_test.v
+++ b/vlib/v/tests/interface_str_method_test.v
@@ -18,3 +18,18 @@ fn test_interface_str_method() {
 	ret := printer(s)
 	assert ret == 's'
 }
+
+// for test interface gen to string
+interface Abc {}
+
+struct Xyz {}
+
+fn test_interface_gen_to_string() {
+	d := Abc(Xyz{})
+	mut res := ''
+	if d is Xyz {
+		println(d)
+		res = '${d}'
+	}
+	assert res == '&Xyz{}'
+}

